pkg/fill/service.go

```diff
@@ -0,0 +1,185 @@
+		time.Sleep(time.Duration(s.interval) * time.Second)
+
+		// Get watched addresses from the db
+		rows := s.fetchWatchedAddresses()
+
+		// Get the block number to start fill at
+		// Get the block number to end fill at
+		fillWatchedAddresses, minStartBlock, maxEndBlock := s.GetFillAddresses(rows)
+
+		if len(fillWatchedAddresses) > 0 {
+			log.Infof("running watched address gap filler for block range: (%d, %d)", minStartBlock, maxEndBlock)
+		}
+
+		// Fill the missing diffs
+		for blockNumber := minStartBlock; blockNumber <= maxEndBlock; blockNumber++ {
+			params := statediff.Params{
+				IntermediateStateNodes:   true,
+				IntermediateStorageNodes: true,
+				IncludeBlock:             true,
+				IncludeReceipts:          true,
+				IncludeTD:                true,
+				IncludeCode:              true,
+			}
+
+			fillAddresses := []interface{}{}
+			for _, fillWatchedAddress := range fillWatchedAddresses {
+				if blockNumber >= fillWatchedAddress.StartBlock && blockNumber <= fillWatchedAddress.EndBlock {
+					params.WatchedAddresses = append(params.WatchedAddresses, common.HexToAddress(fillWatchedAddress.Address))
+					fillAddresses = append(fillAddresses, fillWatchedAddress.Address)
+				}
+			}
+
+			if len(fillAddresses) > 0 {
+				s.writeStateDiffAt(blockNumber, params)
+				s.UpdateLastFilledAt(blockNumber, fillAddresses)
+			}
+		}
+	}
+}
+
+// GetFillAddresses finds the encompassing range to perform fill for the given watched addresses
+// it also sets the address specific fill range
+func (s *Service) GetFillAddresses(rows []WatchedAddress) ([]WatchedAddress, uint64, uint64) {
+	fillWatchedAddresses := []WatchedAddress{}
+	minStartBlock := uint64(math.MaxUint64)
+	maxEndBlock := uint64(0)
+
+	for _, row := range rows {
+		// Check for a gap between created_at and watched_at
+		// CreatedAt and WatchedAt being equal is considered a gap of one block
+		if row.CreatedAt > row.WatchedAt {
+			continue
+		}
+
+		startBlock := uint64(0)
+		endBlock := uint64(0)
+
+		// Check if some of the gap was filled earlier
+		if row.LastFilledAt > 0 {
+			if row.LastFilledAt < row.WatchedAt {
+				startBlock = row.LastFilledAt + 1
+			}
+		} else {
+			startBlock = row.CreatedAt
+		}
+
+		// Add the address for filling
+		if startBlock > 0 {
+			row.StartBlock = startBlock
+			if startBlock < minStartBlock {
+				minStartBlock = startBlock
+			}
+
+			endBlock = row.WatchedAt
+			row.EndBlock = endBlock
+			if endBlock > maxEndBlock {
+				maxEndBlock = endBlock
+			}
+
+			fillWatchedAddresses = append(fillWatchedAddresses, row)
+		}
+	}
+
+	return fillWatchedAddresses, minStartBlock, maxEndBlock
+}
+
+// UpdateLastFilledAt updates the fill status for the provided addresses in the db
+func (s *Service) UpdateLastFilledAt(blockNumber uint64, fillAddresses []interface{}) {
+	// Prepare the query
+	query := "UPDATE eth_meta.watched_addresses SET last_filled_at=? WHERE address IN (?" + strings.Repeat(",?", len(fillAddresses)-1) + ")"
```

Comment:
ipld-eth-server writing to the DB itself is outside of its usual scope

Reply:
Implemented as per the discussion here: https://github.com/vulcanize/go-ethereum/issues/127#issuecomment-1011819114
We can move the service to a new repository if desired.

Reply:
Not sure I was part of that discussion

Reply:
Curious what @ashwinphatak  thinks, I just want to be cautious about adding more and more misc bells and whistles to this repo as the original intent was for it to have a focused and well-defined function that was encapsulated by its name.

Reply:
And the clear delineation that this binary was read-only on the DB is kind of nice from an ops perspective.

Reply:
But on the otherhand, Shane not having to keep track of another binary has its benefits from the ops perspective.

Reply:
The watched addresses fill service will be extracted to a separate repository.
See: https://github.com/vulcanize/ops/issues/156